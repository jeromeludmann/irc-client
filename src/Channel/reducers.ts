import { combineReducers } from "redux";
import {
  CommandInputState,
  commandInputReducer,
} from "@app/CommandInput/reducers";
import { ChannelScope } from "@app/types";

// Name

type NameState = string;
const nameInitialState: NameState = "";
const nameReducer = (state = nameInitialState): NameState => state;

// Messages

type MessagesState = string[];

type MessagesPayload = ChannelScope & { value: string };

interface MessagesAction {
  type: "COMMAND_SENT";
  payload: MessagesPayload;
}

const messagesInitialState: MessagesState = [];

const messagesReducer = (
  state = messagesInitialState,
  { type, payload }: MessagesAction,
): MessagesState => {
  switch (type) {
    case "COMMAND_SENT":
      return [...state, payload.value];
    default:
      return state;
  }
};

export interface ChannelState {
  name: NameState;
  messages: MessagesState;
  commandInput: CommandInputState;
}

export const channelReducer = combineReducers<ChannelState>({
  name: nameReducer,
  messages: messagesReducer,
  commandInput: commandInputReducer,
});
