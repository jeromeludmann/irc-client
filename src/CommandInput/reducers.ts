import { combineReducers } from "redux";

// Value

type ValueState = string;

type ValueAction =
  | {
      type: "COMMAND_CHANGED";
      payload: { value: string };
    }
  | {
      type: "COMMAND_SENT";
      payload: {};
    };

const valueInitialState: ValueState = "";

const value = (state = valueInitialState, action: ValueAction): ValueState => {
  switch (action.type) {
    case "COMMAND_CHANGED":
      return action.payload.value;
    case "COMMAND_SENT":
      return "";
    default:
      return state;
  }
};

// History

type HistoryState = string[];
const historyInitialState: HistoryState = [];
const history = (state = historyInitialState): HistoryState => state;

// CommandInput

export type CommandInputState = {
  value: ValueState;
  history: HistoryState;
};

export const commandInputReducer = combineReducers<CommandInputState>({
  value,
  history,
});
