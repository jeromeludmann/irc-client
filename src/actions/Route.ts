import { Action } from "redux";

export interface Route {
  server: string;
  channel: string;
}

export type RoutedAction = Action & {
  route: Route;
};
