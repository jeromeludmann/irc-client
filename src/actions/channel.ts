import { ChannelScope } from "@app/types";

export const ChannelListTypes = {
  ADD: "CHANNEL/ADD",
  REMOVE: "CHANNEL/REMOVE",
};

export interface AddChannelAction {
  type: typeof ChannelListTypes.ADD;
  payload: ChannelScope;
}

export interface RemoveChannelAction {
  type: typeof ChannelListTypes.REMOVE;
  payload: ChannelScope;
}
