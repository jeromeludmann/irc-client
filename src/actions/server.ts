import { Action } from "redux";

export const CONNECT_SERVER = "SERVER/CONNECT_SERVER";
export const DISCONNECT_SERVER = "SERVER/DISCONNECT_SERVER";

export type ConnectServerAction = Action<typeof CONNECT_SERVER> & {
  payload: {
    host: string;
    port: number;
  };
};

export type DisconnectServerAction = Action<typeof DISCONNECT_SERVER>;

export const connectServer = (
  host: string,
  port: number = 6667,
): ConnectServerAction => ({
  type: CONNECT_SERVER,
  payload: { host, port },
});

export const disconnectServer = (): DisconnectServerAction => ({
  type: DISCONNECT_SERVER,
});
