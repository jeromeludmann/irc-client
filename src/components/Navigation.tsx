import React, { PureComponent, MouseEvent, CSSProperties } from "react";

interface Props {
  servers: {
    [s: string]: {
      windows: { [w: string]: { unread: boolean } };
    };
  };
  active: {
    server: string;
    window: string;
  };
  onWindowButtonClick: (
    { server, window }: { server: string; window: string },
  ) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { active, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(server => (
          <ul key={server} style={{ listStyle: "none", padding: "0" }}>
            {Object.keys(servers[server].windows).map(window => {
              const isActive =
                server === active.server && window === active.window;
              const isUnread = servers[server].windows[window].unread;
              const style: CSSProperties = {
                outline: "none",
                border: "1px solid #ddd",
                width: "100px",
                color: isActive ? "black" : "#bbb",
                backgroundColor: isUnread ? "yellow" : "#fff",
              };

              return (
                <li key={window} style={{ marginBottom: "5px" }}>
                  <button
                    onClick={this.handleClick}
                    data-server={server}
                    data-window={window}
                    style={style}
                  >
                    {window}
                  </button>
                </li>
              );
            })}
          </ul>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const server = e.currentTarget.getAttribute("data-server");
    const window = e.currentTarget.getAttribute("data-window");

    if (server && window) {
      this.props.onWindowButtonClick({ server, window });
    }
  };
}
