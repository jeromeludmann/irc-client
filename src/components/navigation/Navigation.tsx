import React, { PureComponent, MouseEvent } from "react";
import { STATUS, RAW } from "@app/Route";
import { List, ListItem, Button } from "@app/components/navigation/style";

interface Props {
  servers: {
    [serverKey: string]: {
      name: string;
      channels: { [channelKey: string]: { activity: boolean } };
    };
  };
  window: {
    serverKey: string;
    channelKey: string;
  };
  onChannelButtonClick: (
    { serverKey, channelKey }: { serverKey: string; channelKey: string }
  ) => void;
}

export default class Navigation extends PureComponent<Props> {
  public render() {
    const { window, servers } = this.props;

    return (
      <>
        {Object.keys(servers).map(serverKey => (
          <List key={serverKey}>
            {Object.keys(servers[serverKey].channels).map(channelKey => (
              <ListItem key={channelKey}>
                <Button
                  onClick={this.handleClick}
                  data-server-key={serverKey}
                  data-channel-key={channelKey}
                  isActive={
                    serverKey === window.serverKey &&
                    channelKey === window.channelKey
                  }
                  hasActivity={servers[serverKey].channels[channelKey].activity}
                >
                  {channelKey === STATUS
                    ? servers[serverKey].name
                    : channelKey === RAW
                      ? serverKey
                      : channelKey}
                </Button>
              </ListItem>
            ))}
          </List>
        ))}
      </>
    );
  }

  private handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const serverKey = e.currentTarget.getAttribute("data-server-key") as string;
    const channelKey = e.currentTarget.getAttribute(
      "data-channel-key"
    ) as string;

    this.props.onChannelButtonClick({ serverKey, channelKey });
  };
}
