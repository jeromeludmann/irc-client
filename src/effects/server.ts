import { Socket } from 'net'
import { eventChannel, END } from 'redux-saga'
import { takeEvery, call, put, take, select } from 'redux-saga/effects'
import {
  CONNECT_TO_SERVER,
  SEND_RAW_MESSAGE,
  DISCONNECT_FROM_SERVER,
  ConnectToServerAction,
  lookup,
  setConnectionEstablished,
  receiveRawMessages,
  setConnectionClosed,
  setConnectionFailed,
  SendRawMessageAction,
  DisconnectFromServerAction,
} from '@app/actions/socket'
import { sendQuit } from '@app/actions/messages/outgoing'
import { CRLF } from '@app/utils/helpers'
import { addNewServer } from '@app/actions/ui'
import { BufferKey } from '@app/utils/Route'
import { getServerKeys } from '@app/state/root/selectors'
import { connect, remove, send, end } from './socket'

export function* server() {
  yield takeEvery(CONNECT_TO_SERVER, connectToServer)
  yield takeEvery(SEND_RAW_MESSAGE, sendRawMessage)
  yield takeEvery(DISCONNECT_FROM_SERVER, disconnectFromServer)
}

function* connectToServer(action: ConnectToServerAction) {
  console.log('connectToServer: started', action)

  const {
    payload: { host, port, newConnection },
  } = action

  const serverKey = newConnection
    ? yield generateServerKey()
    : action.route.serverKey

  const socket: Socket = yield* connect(
    serverKey,
    host,
    port,
  )

  const socketChannel = yield call(createSocketChannel, serverKey, socket)

  try {
    while (true) {
      const socketAction = yield take(socketChannel)
      yield put(socketAction)
    }
  } finally {
    yield* remove(serverKey)

    console.log('connectToServer: ended', action)
  }
}

function* sendRawMessage(action: SendRawMessageAction) {
  console.log('sendRawMessage: started', action)

  yield* send(action.route.serverKey, action.payload.raw)

  if (action.embeddedAction !== undefined) {
    yield put(action.embeddedAction)
  }

  console.log('sendRawMessage: ended', action)
}

function* disconnectFromServer(action: DisconnectFromServerAction) {
  console.log('disconnectFromServer: started', action)

  yield put(sendQuit(action.route.serverKey, action.payload.quitMessage))
  yield* end(action.route.serverKey)

  console.log('disconnectFromServer: ended', action)
}

function* generateServerKey() {
  const existingServerKeys: string[] = yield select(getServerKeys)

  const serverKey: string = yield call(function generateKey(): string {
    const key = Math.random()
      .toString(36)
      .slice(2)
    return existingServerKeys.includes(key) ? generateKey() : key
  })

  yield put(addNewServer({ serverKey, bufferKey: BufferKey.NONE }))
  return serverKey
}

function createSocketChannel(serverKey: string, socket: Socket) {
  return eventChannel(emit => {
    let buffer = ''

    socket.on('lookup', (error, address, family, host) => {
      emit(lookup(serverKey, error, address, family, host))
    })

    socket.on('connect', () => {
      emit(setConnectionEstablished(serverKey))
    })

    socket.on('data', data => {
      buffer += data
      const messages = buffer.split(CRLF)
      buffer = messages.pop() || ''

      emit(receiveRawMessages(serverKey, messages))
    })

    socket.on('close', hadError => {
      emit(setConnectionClosed(serverKey, hadError))
      emit(END)
    })

    socket.on('error', ({ name, message, stack }) => {
      emit(setConnectionFailed(serverKey, name, message, stack))
      emit(END)
    })

    socket.on('end', () => {
      console.warn('unhandled socket end')
    })

    socket.on('timeout', () => {
      console.warn('unhandled socket timeout')
    })

    return () => {
      console.log('end of socket channel', serverKey)
    }
  })
}
