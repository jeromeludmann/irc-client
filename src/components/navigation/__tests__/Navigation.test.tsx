import React from "react";
import { shallow, mount } from "enzyme";
import Navigation from "@app/components/navigation/Navigation";
import { STATUS } from "@app/Route";
import { channelsInitialState } from "@app/reducers/server/channels";
import { channelInitialState } from "@app/reducers/channel";

describe("Navigation component", () => {
  const props = {
    onChannelButtonClick: jest.fn(),
    servers: {
      server1: {
        name: "server.domain",
        channels: {
          ...channelsInitialState,
          "#channel": { ...channelInitialState, activity: true }
        }
      }
    },
    window: { serverKey: "server1", channelKey: STATUS }
  };

  it("should render correctly", () => {
    expect(shallow(<Navigation {...props} />)).toMatchSnapshot();
  });

  it("should handle button click event", () => {
    mount(<Navigation {...props} />)
      .find("button")
      .first()
      .simulate("click");
  });
});
