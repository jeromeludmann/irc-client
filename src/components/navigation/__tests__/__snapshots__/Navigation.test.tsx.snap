// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navigation
    onChannelButtonClick={[MockFunction]}
    servers={
      Object {
        "server1": Object {
          "channels": Object {
            "#channel": Object {
              "activity": true,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 2,
                  "values": Array [
                    "/connect -n localhost",
                    "/connect localhost",
                  ],
                },
                "value": "",
              },
              "messages": Array [],
            },
            "@raw": Object {
              "activity": false,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 2,
                  "values": Array [
                    "/connect -n localhost",
                    "/connect localhost",
                  ],
                },
                "value": "",
              },
              "messages": Array [],
            },
            "@status": Object {
              "activity": false,
              "input": Object {
                "dirtyValue": "",
                "history": Object {
                  "index": 2,
                  "values": Array [
                    "/connect -n localhost",
                    "/connect localhost",
                  ],
                },
                "value": "",
              },
              "messages": Array [],
            },
          },
          "name": "server.domain",
        },
      }
    }
    window={
      Object {
        "channelKey": "@status",
        "serverKey": "server1",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <styled.ul>
          <styled.li>
            <styled.button
              data-channel-key="@raw"
              data-server-key="server1"
              hasActivity={false}
              isActive={false}
              onClick={[Function]}
            >
              server1
            </styled.button>
          </styled.li>
          <styled.li>
            <styled.button
              data-channel-key="@status"
              data-server-key="server1"
              hasActivity={false}
              isActive={true}
              onClick={[Function]}
            >
              server.domain
            </styled.button>
          </styled.li>
          <styled.li>
            <styled.button
              data-channel-key="#channel"
              data-server-key="server1"
              hasActivity={true}
              isActive={false}
              onClick={[Function]}
            >
              #channel
            </styled.button>
          </styled.li>
        </styled.ul>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "server1",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.li>
              <styled.button
                data-channel-key="@raw"
                data-server-key="server1"
                hasActivity={false}
                isActive={false}
                onClick={[Function]}
              >
                server1
              </styled.button>
            </styled.li>,
            <styled.li>
              <styled.button
                data-channel-key="@status"
                data-server-key="server1"
                hasActivity={false}
                isActive={true}
                onClick={[Function]}
              >
                server.domain
              </styled.button>
            </styled.li>,
            <styled.li>
              <styled.button
                data-channel-key="#channel"
                data-server-key="server1"
                hasActivity={true}
                isActive={false}
                onClick={[Function]}
              >
                #channel
              </styled.button>
            </styled.li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "@raw",
            "nodeType": "class",
            "props": Object {
              "children": <styled.button
                data-channel-key="@raw"
                data-server-key="server1"
                hasActivity={false}
                isActive={false}
                onClick={[Function]}
              >
                server1
              </styled.button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "server1",
                "data-channel-key": "@raw",
                "data-server-key": "server1",
                "hasActivity": false,
                "isActive": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "server1",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "@status",
            "nodeType": "class",
            "props": Object {
              "children": <styled.button
                data-channel-key="@status"
                data-server-key="server1"
                hasActivity={false}
                isActive={true}
                onClick={[Function]}
              >
                server.domain
              </styled.button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "server.domain",
                "data-channel-key": "@status",
                "data-server-key": "server1",
                "hasActivity": false,
                "isActive": true,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "server.domain",
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "#channel",
            "nodeType": "class",
            "props": Object {
              "children": <styled.button
                data-channel-key="#channel"
                data-server-key="server1"
                hasActivity={true}
                isActive={false}
                onClick={[Function]}
              >
                #channel
              </styled.button>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "#channel",
                "data-channel-key": "#channel",
                "data-server-key": "server1",
                "hasActivity": true,
                "isActive": false,
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "#channel",
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <styled.ul>
            <styled.li>
              <styled.button
                data-channel-key="@raw"
                data-server-key="server1"
                hasActivity={false}
                isActive={false}
                onClick={[Function]}
              >
                server1
              </styled.button>
            </styled.li>
            <styled.li>
              <styled.button
                data-channel-key="@status"
                data-server-key="server1"
                hasActivity={false}
                isActive={true}
                onClick={[Function]}
              >
                server.domain
              </styled.button>
            </styled.li>
            <styled.li>
              <styled.button
                data-channel-key="#channel"
                data-server-key="server1"
                hasActivity={true}
                isActive={false}
                onClick={[Function]}
              >
                #channel
              </styled.button>
            </styled.li>
          </styled.ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "server1",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <styled.li>
                <styled.button
                  data-channel-key="@raw"
                  data-server-key="server1"
                  hasActivity={false}
                  isActive={false}
                  onClick={[Function]}
                >
                  server1
                </styled.button>
              </styled.li>,
              <styled.li>
                <styled.button
                  data-channel-key="@status"
                  data-server-key="server1"
                  hasActivity={false}
                  isActive={true}
                  onClick={[Function]}
                >
                  server.domain
                </styled.button>
              </styled.li>,
              <styled.li>
                <styled.button
                  data-channel-key="#channel"
                  data-server-key="server1"
                  hasActivity={true}
                  isActive={false}
                  onClick={[Function]}
                >
                  #channel
                </styled.button>
              </styled.li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "@raw",
              "nodeType": "class",
              "props": Object {
                "children": <styled.button
                  data-channel-key="@raw"
                  data-server-key="server1"
                  hasActivity={false}
                  isActive={false}
                  onClick={[Function]}
                >
                  server1
                </styled.button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "server1",
                  "data-channel-key": "@raw",
                  "data-server-key": "server1",
                  "hasActivity": false,
                  "isActive": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "server1",
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "@status",
              "nodeType": "class",
              "props": Object {
                "children": <styled.button
                  data-channel-key="@status"
                  data-server-key="server1"
                  hasActivity={false}
                  isActive={true}
                  onClick={[Function]}
                >
                  server.domain
                </styled.button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "server.domain",
                  "data-channel-key": "@status",
                  "data-server-key": "server1",
                  "hasActivity": false,
                  "isActive": true,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "server.domain",
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "#channel",
              "nodeType": "class",
              "props": Object {
                "children": <styled.button
                  data-channel-key="#channel"
                  data-server-key="server1"
                  hasActivity={true}
                  isActive={false}
                  onClick={[Function]}
                >
                  #channel
                </styled.button>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "#channel",
                  "data-channel-key": "#channel",
                  "data-server-key": "server1",
                  "hasActivity": true,
                  "isActive": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "#channel",
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
